File: leetcode/90.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
  vector<vector<int>> subsetsWithDup(vector<int> &nums) {
    vector<vector<int>> ans(1, vector<int>());
    unordered_map<int, int> help;
    int n = nums.size();
    sort(nums.begin(), nums.end());
    for (int i = 0; i < n; i++) {
      int val = nums[i];
      int len = ans.size();
      int j = help.find(val) == help.end() ? 0 : help[val];
      for (; j < len; j++) {
        vector<int> cur = ans[j];
        cur.emplace_back(val);
        ans.emplace_back(cur);
      }
      help[val] = len;
    }
    return ans;
  }
};

int main() {
  vector<int> nums = {1, 2, 2};
  Solution().subsetsWithDup(nums);
  return 0;
}

File: leetcode/90.py
from typing import List
import collections
import copy


class Solution:

    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = [[]]
        help = collections.defaultdict(int)
        for i in range(len(nums)):
            j = help[nums[i]]
            n = len(ans)
            while j < n:
                cur = copy.deepcopy(ans[j])
                cur.append(nums[i])
                ans.append(cur)
                j += 1
            help[nums[i]] = j
        return ans


if __name__ == "__main__":
    print(Solution().subsetsWithDup([1, 2, 2]))
